packages/styled-components/src/models/StyledComponent.ts

```diff
@@ -29,24 +31,40 @@ import isTag from '../utils/isTag';
 import joinStrings from '../utils/joinStrings';
 import merge from '../utils/mixinDeep';
 import ComponentStyle from './ComponentStyle';
+import Keyframes from './Keyframes';
 import { useStyleSheet, useStylis } from './StyleSheetManager';
 import { ThemeContext } from './ThemeProvider';
 
-const identifiers: { [key: string]: number } = {};
+/** Shared context for component identifiers, preventing duplicate ids */
+const COMPONENT_IDENTIFIERS: Record<string, Record<string, number>> = {}
+
 
 /* We depend on components having unique IDs */
-function generateId(displayName?: string, parentComponentId?: string): string {
-  const name = typeof displayName !== 'string' ? 'sc' : escape(displayName);
-  // Ensure that no displayName can lead to duplicate componentIds
-  identifiers[name] = (identifiers[name] || 0) + 1;
+function generateId<OuterProps extends object>(rules: RuleSet<OuterProps>, displayName?: string, parentComponentId?: string): string {
+  const namespace = typeof displayName !== 'string' ? 'sc' : escape(displayName);
+
+  /** Keyframes cannot be interpolated without stylesheet */
+  const isNotKeyframe = (interpolation: Interpolation<OuterProps>) => !(interpolation instanceof Keyframes)
+  const styleSheet = flatten(rules).filter(isNotKeyframe).join('')
```

Comment:
Some tests were failing due to "_It seems you are interpolating a keyframe declaration (%s) into an untagged string_"; at first glance I couldn't find a way to use the `css` helper here, so I chose to filter any keyframes out from the hashing. If all the other CSS rules would be identical between two components, the `increment` number would handle it.

Reply:
Hmm, maybe using `JSON.stringify()` here would produce similar output while being faster.

Reply:
This filtering is no longer necessary with `JSON.stringify()`.